Move python-rhsm build into subscription-manager

kahowell opened the pull request:
Most invasive changes were:
 - Moved most of the python files to sitearch (/usr/lib64 instead of
   /usr/lib), since that is where setuptools installs them due to
   building the certificate extension in python-rhsm
 - Changes to get setup.py to be python3 compatible
 - Excluded some Python 3 artifacts in the spec file for now, since we
   are not yet building subscription-manager for Python 3.
 - Moved tests for python-rhsm from `python-rhsm/test/` to `test/rhsm`.
 - Removed the python-rhsm README.

This will require some changes to our jobs in order to work correctly.
I'll open a PR against candlepin-jobs soon.

kahowell commented:
Needs candlepin/candlepin-jobs#22 first, so that CI tests can be adjusted appropriately.

kahowell commented:
6c20b1a..230665e: rebase, remove a few dangling files in python-rhsm

kahowell commented:
230665e..3ae0482: default to `python` instead of `/usr/bin/python` in Makefile. Fixes lots of various issues that cropped up in python dependencies

kahowell commented:
3ae0482..5596d15:
 - don't run functional tests
 - fix stylish issues in rhsm tests

kahowell commented:
5596d15..193a66f: install etc-conf/ca/redhat-uep.pem explicitly

kahowell commented:
193a66f..1be8914: install pip deps with `-I` which means nose gets installed in the virtualenv even if it's on the system already

Needed this to work around a test failure...

awood commented on subscription-manager.spec:
```diff
@@ -290,6 +302,61 @@ the remote in the currently deployed .origin file.
 %endif
 
 
+%package -n python-rhsm
```
Have we ruled out renaming the package to `subscription-manager-rhsm` or the like and using the `Obsoletes` and `Provides` directives in the package header?

kahowell replied:
> Have we ruled out renaming the package to subscription-manager-rhsm or the like and using the Obsoletes and Provides directives in the package header?

No, we have not ruled this out. What are your thoughts on benefits of doing that?

awood replied:
I think the main benefit is it allows people to find the source.  If I see a package installed on my system named "python-rhsm", I'm going to try running `fedpkg clone python-rhsm`.  That won't work and then I have to go to Koji and figure out where the heck "python-rhsm" comes from.  In my opinion, renaming the package to "subscription-manager-rhsm" makes it apparent where the code is coming from.  It's a nice-to-have I think.

Renaming is possible.  Fedora has instructions [here](https://fedoraproject.org/wiki/Packaging:Guidelines#Renaming.2FReplacing_Existing_Packages).

jirihnidek replied:
I like this renaming too! 👍

awood commented on subscription-manager.spec:
```diff
@@ -359,18 +438,21 @@ rm -rf %{buildroot}
-%dir %{python_sitelib}/subscription_manager/branding
-%dir %{python_sitelib}/subscription_manager/model
-%dir %{python_sitelib}/subscription_manager/plugin
+%dir %{python_sitearch}/rhsmlib/candlepin
```
Are we actually required to move to sitearch?  The rhsm subpackage will need to go there, but the subscription-manager stuff doesn't have to does it?  I don't oppose the move, but I just want to make sure we considered both options.

kahowell replied:
Unfortunately, if we build it as one distribution (one `setup.py`, then everything ends up in `sitearch` due to logic that [puts any distribution that has an extension into a platform-specific directory](https://github.com/python/cpython/blob/b7cbfe49e39a7bbd7da20b937735a8a60bbf1872/Lib/distutils/command/install.py#L336).

awood replied:
Interesting!  No big deal.  Just thought I'd ask.  We'll just have a really big RPMDiff section to waive :)

jirihnidek replied:
👍 for having all modules in `/usr/lib64/pythonX.Y/site-packages`

awood commented on test-requirements.txt:
```diff
@@ -12,5 +12,7 @@ freezegun
 mock
 Sphinx
-# git+https://github.com/alikins/pyqver.git#egg=pyqver
+yanc
```
I would nix this as a test requirement.  It just makes the output colorful and it has some [problems](https://github.com/0compute/yanc/issues/7) (the author claims that issue has been fixed in nose, but I still hit it when I was using yanc..

If a developer wants it, they can always `pip` install it themselves, but it's not really a requirement.  Same with `xtraceback`.  And we'd need to also remove the stuff in the `.noserc`/`setup.cfg` that references these plugins

kahowell commented:
b93dab57b..04a7819cf: rebase

@awood, @jirihnidek, thanks for the reviews. I've added a few more commits, and rebased against master. Please take another look when you get a chance.

I still need to squash once we're good with the changes.

jirihnidek commented on subscription-manager.spec:
```diff
@@ -290,6 +302,61 @@ the remote in the currently deployed .origin file.
+Requires: python-rhsm-certificates = %{version}-%{release}
+
+# Required by Fedora packaging guidelines
+%{?python_provide:%python_provide python3-python-rhsm}
```
It works in master, but it didn't work in your branch. I can confirm that it is fixed now.

awood commented on src/subscription_manager/utils.py:
```diff
@@ -377,12 +377,12 @@ def set_filter_string(self, filter_string):
+                (?:\\\\.?)*       # Anything escaped with a backslash, or just a trailing backslash
             )*)                 # Repeat the above sequence 0+ times, greedily
-            ([*?]|\Z)           # Any of our wildcards (* or ?) not preceded by a backslash OR end of input
+            ([*?]|\\Z)           # Any of our wildcards (* or ?) not preceded by a backslash OR end of input
```
Were these regexs wrong?  As I understand it, the new regex will match "\Z" instead of end of input.

kahowell replied:
> Were these regexs wrong? As I understand it, the new regex will match "\Z" instead of end of input.

Don't think so. Reason for this change was that raw unicode literals were removed from Python 3... and when running setup.py, we end up importing or otherwise processing `utils.py` (don't recall specifics)... and so I needed to change the literal to a plain unicode literal... This meant double escaping.

You can do this to sanity-check though: take both representations and confirm that they produce equivalent strings:

```python
expression1 = ur"""
            ((?:                # A captured, non-capture group :)
                [^*?\\]*        # Character literals and other uninteresting junk (greedy)
                (?:\\.?)*       # Anything escaped with a backslash, or just a trailing backslash
            )*)                 # Repeat the above sequence 0+ times, greedily
            ([*?]|\Z)           # Any of our wildcards (* or ?) not preceded by a backslash OR end of input
        """

expression2 = u"""
            ((?:                # A captured, non-capture group :)
                [^*?\\\\]*        # Character literals and other uninteresting junk (greedy)
                (?:\\\\.?)*       # Anything escaped with a backslash, or just a trailing backslash
            )*)                 # Repeat the above sequence 0+ times, greedily
            ([*?]|\\Z)           # Any of our wildcards (* or ?) not preceded by a backslash OR end of input
        """

print(expression1 == expression2)
```

outputs `True`

kahowell commented:
04a7819cf..af0187a98: rebase against master

kahowell commented:
bd1088ea3..f445b6486: squashed
